Add Quadrature Decoder support for S32

sumitbatra-nxp opened the pull request:
This driver configures and programs Lcu, Trgmux and Emios_Icu modules to get the rotations reported by the quadrature encoder in radians 
Testing Hardware: CANHUBK3 board

We have raised following new PR in hal_nxp to support this code - 
https://github.com/zephyrproject-rtos/hal_nxp/pull/297
and 
https://github.com/zephyrproject-rtos/hal_nxp/pull/306

manuargue commented on dts/bindings/sensor/nxp,s32k-qdec.yaml:
```diff
@@ -0,0 +1,112 @@
+    description: |
+      This gives the hardware instance of LCU module.
+
+  emios_inst:
```
The common pattern is to have a phandle to the eMIOS device (e.g. `nxp,emios`) and use a device API to retrieve the hw instance number. The same applies for `lcu_inst` and `trgmux-inst` properties above.
There is already a generic eMIOS driver (https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/misc/nxp_s32_emios/nxp_s32_emios.c) that could be extended for that purpose.

sumitbatra-nxp replied:
If I understand these 2 drivers - nxp_s32_emios.c and pwm_nxp_s32_emios.c drivers correctly then, to be able to extract the handle of eMIOS your driver's DT node should be correlated to the eMIOS DT node which has the REG ADDR of an eMIOS instance which can be compared with the IP_EMIOS_##idx##_BASE in HAL and added in a list using LISTIFY.

But my DT node is a QDEC node which has LCU, TRGMUX and eMIOS, which can't be a direct child of eMIOS.
But I believe we can separate out these drivers and DT nodes.. we can discuss about this in detail

Dat-NguyenDuy replied:
Yep, but i think is also relates to what Manuel suggested to breakdown the current qdec driver.(https://github.com/zephyrproject-rtos/zephyr/pull/65035#pullrequestreview-1723740577)

sumitbatra-nxp replied:
fixed for eMIOS.. but for lcu and trgmux I would need separate DT nodes for referring to their phandles, which will be possible only once I separate them out

sumitbatra-nxp replied:
Added phandles for lcu and trgmux

manuargue commented on drivers/sensor/qdec_s32k/qdec_s32k.c:
```diff
@@ -0,0 +1,448 @@
+/* LCU Logic Input Configuration */
+#define LogicInputCfg_Common(MuxSel)								   \
+	{											   \
+		/* uint8 MuxSel		*/ MuxSel,						   \
```
Instead of the comment `/* uint8 MuxSel */`, use the member name as `.MuxSel = MuxSel`. The S32 HAL is conditionally compiled and some members may not exist for some devices, so relying in the order is fragile. And also is self-documenting.

sumitbatra-nxp replied:
fixed!

manuargue commented on dts/bindings/sensor/nxp,s32k-qdec.yaml:
```diff
@@ -0,0 +1,112 @@
+      This is the array containing 2 emios channel TypeG numbers used by the qdec.
+      See Table 342 in S32K3XXRM for more information
+
+  trgmux-maps:
```
If see this is used to configure pinmuxing of eMIOS or TRGMUX peripheral: https://github.com/zephyrproject-rtos/zephyr/blob/a53e882ff3777a3c67c03da984106a4625f56ebb/drivers/sensor/qdec_s32k/qdec_s32k.c#L151
There's already a pinctrl driver for that purpose, please add the needed pinmuxing configuration using `pinctrl` node in the board's overlay, and remove this property.

sumitbatra-nxp replied:
I'll need to add new pinmux entries in - hal_nxp/dts/nxp/s32/S32/S32K344-172MQFP-pinctrl.h 
I hope that's ok ?

sumitbatra-nxp replied:
fixed !

sumitbatra-nxp replied:
fixed !

manuargue replied:
The binding property and board configuration is still there, https://github.com/zephyrproject-rtos/zephyr/blob/2a924701eda579fb9c526edab42d1f37478c90cd/boards/arm/mr_canhubk3/mr_canhubk3.dts#L94-L96 pls double check

sumitbatra-nxp replied:
Changed as per suggestions !

manuargue reviewed (changes requested):
On a first look to this pr, IMO the drivers for the peripherals TRGMUX, LCU and eMIOS needs to be further split into individual drivers so that the QDEC sensor driver integrates them together to perform the quadrature specific functionality.

Particularly eMIOS is a versatile peripheral that has multiple uses. We have already supported PWM based on eMIOS, so there is already some generic functionality/initialization for the eMIOS module itself abstracted. This should be further extended to support your use-case. Another reference of this pattern that you can find in tree, is the QuadSPI memory controller (memc) and the flash-technology-specific QuadSPI NOR driver (flash) using the S32 QSPI controller.

manuargue commented:
@Dat-NguyenDuy could you please have a review on this PR? mostly related to how this will affect the existing eMIOS/PWM support.

(btw, @Dat-NguyenDuy is added as collaborators in maintainers file but still cannot be added as reviewer, we are checking on that).

manuargue commented:
> We have raised following new PR in hal_nxp to support this code -
https://github.com/zephyrproject-rtos/hal_nxp/pull/297

@sumitbatra-nxp pls add `pull/297/head` as revision of `hal_nxp` module in `west.yaml` to link both prs.

sumitbatra-nxp commented on dts/bindings/sensor/nxp,s32k-qdec.yaml:
```diff
@@ -0,0 +1,112 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Quadrature Decoder driver which processes encoder signals to determine MR-BUGGY3 displacement
```
corrected !

manuargue replied:
there are still mentions to S32K3
https://github.com/zephyrproject-rtos/zephyr/blob/2a924701eda579fb9c526edab42d1f37478c90cd/dts/bindings/sensor/nxp%2Cs32-qdec.yaml#L7
anything specific to `mr_canhubk3` should be moved to the boards doc.

sumitbatra-nxp replied:
there is nothing specific to the board.. I missed this instance.. corrected now !

manuargue replied:
pls see https://github.com/zephyrproject-rtos/zephyr/blob/6215845404cb2a399a9fc52c5a6aed9685d26d98/dts/bindings/sensor/nxp%2Cs32-qdec.yaml#L104

sumitbatra-nxp replied:
removed !

sumitbatra-nxp commented on dts/bindings/sensor/nxp,s32k-qdec.yaml:
```diff
@@ -0,0 +1,112 @@
+        emios-channels = <6 7>;
+   };
+
+compatible: "nxp,qdec-s32k"
```
corrected in both hal and zephyr

Dat-NguyenDuy commented on drivers/sensor/qdec_s32/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# Copyright 2023 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config QDEC_S32
```
Not much important but I would suggest to use `NXP_S32` instead of `S32` in the name of at least Kconfig and files, just to avoid ambiguity and consistent with the rest of NXP S32 drivers

sumitbatra-nxp replied:
Changed !

Dat-NguyenDuy commented on drivers/sensor/qdec_s32/qdec_s32.c:
```diff
@@ -0,0 +1,474 @@
+			EMIOS_ICU_MODE_EDGE_COUNTER, EMIOS_ICU_MODE_WITHOUT_DMA,		\
+			EMIOS_ICU_NO_MEASUREMENT, EMIOS_ICU_RISING_EDGE,			\
+			EMIOS_DIGITAL_FILTER_BYPASSED, NULL_PTR, NULL_PTR, (uint8)255U,		\
+			(boolean)FALSE, .eMiosChannelNotification = NULL_PTR,			\
```
Is this possible to configure overflow callback here? instead of touching on HAL's internal state like you did above: https://github.com/zephyrproject-rtos/zephyr/blob/f2e74229013c15f579854e4be61dfd3da16b18f4/drivers/sensor/qdec_s32/qdec_s32.c#L199-L202

sumitbatra-nxp replied:
@Dat-NguyenDuy .. Assigning anything to these structure members will only be effective if I can pass -  eMios_Icu_Ip_ConfigType **eMios_Icu_Ip_Config** to Emios_Icu_Ip_Init. But since I cannot call it so I need to either call Emios exposed API's or use these global variables.
But as discussed we will fix it in a follow up PR

sumitbatra-nxp commented:
> On a first look to this pr, IMO the drivers for the peripherals TRGMUX, LCU and eMIOS needs to be further split into individual drivers so that the QDEC sensor driver integrates them together to perform the quadrature specific functionality.
> 
> Particularly eMIOS is a versatile peripheral that has multiple uses. We have already supported PWM based on eMIOS, so there is already some generic functionality/initialization for the eMIOS module itself abstracted. This should be further extended to support your use-case. Another reference of this pattern that you can find in tree, is the QuadSPI memory controller (memc) and the flash-technology-specific QuadSPI NOR driver (flash) using the S32 QSPI controller.

@manuargue .. As discussed, due to the urgency for this code. I'll be taking up this driver split up activity in a follow up PR

Dat-NguyenDuy commented:
Just left some minor comments from my side

manuargue commented:
> > On a first look to this pr, IMO the drivers for the peripherals TRGMUX, LCU and eMIOS needs to be further split into individual drivers so that the QDEC sensor driver integrates them together to perform the quadrature specific functionality.
> > Particularly eMIOS is a versatile peripheral that has multiple uses. We have already supported PWM based on eMIOS, so there is already some generic functionality/initialization for the eMIOS module itself abstracted. This should be further extended to support your use-case. Another reference of this pattern that you can find in tree, is the QuadSPI memory controller (memc) and the flash-technology-specific QuadSPI NOR driver (flash) using the S32 QSPI controller.
> 
> @manuargue .. As discussed, due to the urgency for this code. I'll be taking up this driver split up activity in a follow up PR

As far as the NXP S32 integration goes, I'm okay to accept a first version to get the QDEC functionality working provided a github issue is logged to address all improvements mentioned above.

sumitbatra-nxp commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,475 @@
+		eMios_Icu_Ip_ChannelConfig_Common(						\
+			DT_INST_PROP_BY_IDX(n, emios_channels, EMIOS_CCW_CH_IDX)),		\
+	};											\
+	const eMios_Icu_Ip_ConfigType eMios_Icu_Ip_Config = {					\
```
Removed !

sumitbatra-nxp commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,475 @@
+};
+
+struct qdec_s32_data {
+	uint32_t counterCW;
```
Added the snake_case.. will re-run the tests

sumitbatra-nxp commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,475 @@
+	if (ch != SENSOR_CHAN_ALL)
+		return -ENOTSUP;
+
+	static uint32_t prev_time;
```
Removed !

sumitbatra-nxp commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,475 @@
+		.emios_config = &eMios_Icu_Ip_Config,						\
+		.trgmux_config = &Trgmux_Ip_Init_Config,					\
+		.lcu_config = &Lcu_Ip_Init_Config,						\
+		.clock_dev = DEVICE_DT_GET(DT_CLOCKS_CTLR(EMIOS_NXP_S32_NODE(n))),		\
```
As discussed, removed clock_dev and clock_subsys

sumitbatra-nxp commented:
@MaureenHelm .. Can you kindly review.. This feature is urgent for us..

Dat-NguyenDuy commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,424 @@
+
+#define DT_DRV_COMPAT nxp_qdec_s32
+
+#define EMIOS_NXP_S32_NODE(n) DT_NODELABEL(emios##n)
```
pls remove, is not used

MaureenHelm replied:
Still here

sumitbatra-nxp replied:
Removed !

Dat-NguyenDuy commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,424 @@
+#include <Emios_Icu_Ip.h>
+#include <Trgmux_Ip.h>
+#include <Lcu_Ip.h>
+#include <Siul2_Port_Ip.h>
```
seems not used, pls remove

sumitbatra-nxp replied:
Removed the Siul Port include

Dat-NguyenDuy commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,424 @@
+extern eMios_Icu_Ip_ChStateType eMios_Icu_Ip_ChState[EMIOS_ICU_IP_NUM_OF_CHANNELS_USED];
+extern uint8 eMios_Icu_Ip_IndexInChState[EMIOS_ICU_IP_INSTANCE_COUNT][EMIOS_ICU_IP_NUM_OF_CHANNELS];
+
+static uint32_t Clockwise_overflow_count, CounterCW_overflow_count;
```
Perhaps i missed something but look like those are not necessary, covered in here:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/56692056/a4426dcf-83ba-4e55-9d7b-69f47cb7fd86)

sumitbatra-nxp replied:
Removed from qdec_s32_data, as I am not sure how to pass any argument to Emios user notify callback, so using static variables for now.
But will need to add synchronization here, as these variables getting updated in interrupt callback and read in fetch.

sumitbatra-nxp replied:
Added a semaphore

Dat-NguyenDuy replied:
Oh, now i see a semaphore can forever in interrupt context, which shouldn't (https://docs.zephyrproject.org/latest/kernel/services/synchronization/semaphores.html#c.k_sem_take)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/56692056/fcc9b624-ad9c-4e4c-b248-4482a40ba4e6)

I was just asking, if is needed to avoid the race condition, you can justify to keep it

sumitbatra-nxp replied:
@Dat-NguyenDuy  Thanks for pointing it out.. these callbacks are indeed in isr context, so changed the timeout to K_NO_WAIT

MaureenHelm replied:
Don't you need these for each instance? If you can't pass an argument to the callback APIs, then create instance-specific wrapper functions around the common callback function.

MaureenHelm commented on drivers/sensor/qdec_nxp_s32/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+zephyr_library()
+
+zephyr_library_sources_ifdef(CONFIG_QDEC_NXP_S32
```
The ifdef is redundant here. It was already checked in `drivers/sensor/CMakeLists.txt`

sumitbatra-nxp replied:
removed ifdef

MaureenHelm commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,424 @@
+	struct qdec_s32_data *data = dev->data;
+
+	if (ch != SENSOR_CHAN_ALL)
+		return -ENOTSUP;
```
Need to wrap in curly braces

MaureenHelm commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,424 @@
+	};
+
+#define LCU_IP_INIT_CONFIG(n)									\
+	const Lcu_Ip_LogicInputConfigType LogicInput0Cfg =					\
+		LogicInputCfg_Common(LCU_IP_MUX_SEL_LU_IN_0)					\
+	const Lcu_Ip_LogicInputConfigType LogicInput1Cfg =					\
+		LogicInputCfg_Common(LCU_IP_MUX_SEL_LU_IN_1)					\
+	const Lcu_Ip_LogicInputConfigType LogicInput2Cfg =					\
+		LogicInputCfg_Common(LCU_IP_MUX_SEL_LU_OUT_0)					\
+	const Lcu_Ip_LogicInputConfigType LogicInput3Cfg =					\
+		LogicInputCfg_Common(LCU_IP_MUX_SEL_LU_OUT_1)					\
```
Will this work for multiple instances? I don't see the instance number `n` being used here

sumitbatra-nxp replied:
For now we are not planning on using multiple motors, but can think of adding this support later on.
Can we defer it and cover it in - https://github.com/zephyrproject-rtos/zephyr/issues/65459

sumitbatra-nxp replied:
I have added another item, to make this driver multi instance, in the TODO list of #65459

MaureenHelm replied:
No, it needs to be addressed in this PR

MaureenHelm commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,424 @@
+	const Lcu_Ip_LogicInputType LogicInput3_Config =					\
+		LogicInput_Config_Common(n, 3U, LogicInput3Cfg)					\
+												\
+	const Lcu_Ip_LogicInputType								\
+		*const Lcu_Ip_ppxLogicInputArray_Config[LCU_IP_NOF_CFG_LOGIC_INPUTS] = {	\
+			&LogicInput0_Config,							\
+			&LogicInput1_Config,							\
+			&LogicInput2_Config,							\
+			&LogicInput3_Config,							\
+		};										\
```
Will this work for multiple instances? I don't see the instance number `n` being used here

sumitbatra-nxp replied:
Same reply as above !

sumitbatra-nxp replied:
Fixed by the change made for the comment comment

sumitbatra-nxp replied:
@MaureenHelm .. Any S32 based quadrature decoder will have 4 inputs to the LCU, LogicInput0_Config,1,2,3 depict those 4 inputs. 
Now I have made LCU Input Index as a configurable parameter based on device tree selection, there is a logic in hal_nxp which will determine based on this index as to which Logic Cell to be used inside this particular LCU instance. 
Now this makes all the input and output LCU parameters as configurable as per qdec instance

MaureenHelm replied:
This still looks like it won't work for multiple instances because the symbol names (e.g., `LogicInput0Cfg`) aren't unique for each instance. You'll end up with duplicate definitions of the same symbol.

dleach02 replied:
@sumitbatra-nxp, Maureen's point is valid here. You have defined the LCU_IP_INIT_CONFIG() macro that takes an instance. Below it is invoked via the DT__INST_FOREACH which implies you intend for this to work with multiple instances. So when LCU_IP_CONFIG() is invoked multiple times it will try to define the symbols multiple times. If this is indeed intended to support multiple instances, then a symbol like LogicInput0Cfg would need a _##n## to make the name unique.

MaureenHelm commented on dts/bindings/sensor/nxp,s32-qdec.yaml:
```diff
@@ -0,0 +1,104 @@
+
+description: |
+  Quadrature Decoder driver which processes encoder signals to determine motor revs
+
+  with the cooperation of S32 IP blocks- eMIOS, TRGMUX and LCU.
+
+  The sensor qdec application can be used for testing this driver.
+
+  The following example uses TRGMUX IN2 and IN3 to connect to LCU1 LC0 I0 and I1.
+
+  LCU1 LC0 O2 and O3 connect to eMIOS0 CH6(Clockwise rotation) and
+
+  CH7(Counter Clockwise rotation) via TRGMUX_INT_OUT37 and TRGMUX_INT_OUT38
+
+  micro-ticks-per-rev is set as per vehicle gearbox reduction.
+
+  lcu output filters are set to capture maximum speed sensitivity and avoid channel noise.
```
This doesn't need to be double spaced. Remove the empty lines

sumitbatra-nxp replied:
Removed empty lines !

MaureenHelm commented on dts/bindings/sensor/nxp,s32-qdec.yaml:
```diff
@@ -0,0 +1,104 @@
+          <2 TRGMUX_IP_OUTPUT_LCU1_0_INP_I0            TRGMUX_IP_INPUT_SIUL2_IN2>,
+          <3 TRGMUX_IP_OUTPUT_LCU1_0_INP_I1            TRGMUX_IP_INPUT_SIUL2_IN3>;
+
+  lcu_inst:
```
Property names use hyphens instead of underscores by convention

MaureenHelm commented:
Please address the twister failures https://github.com/zephyrproject-rtos/zephyr/runs/18844110050

sumitbatra-nxp commented:
@manuargue @MaureenHelm @Dat-NguyenDuy ..Twister errors have started coming after - 
https://github.com/zephyrproject-rtos/zephyr/pull/62755/
https://github.com/zephyrproject-rtos/zephyr/pull/65173

sumitbatra-nxp commented:
> Please address the twister failures https://github.com/zephyrproject-rtos/zephyr/runs/18844110050

These checks have passed

MaureenHelm commented on dts/bindings/sensor/nxp,s32-qdec.yaml:
```diff
@@ -0,0 +1,96 @@
+      were done based on the current counter's value.
+
+  trgmux-inst:
+    type: int
```
This should be a phandle

sumitbatra-nxp replied:
Currently there is no separate driver and DT node for trgmux and lcu in Zephyr, phandle cannot be used.
Working on separate drivers for trgmux and lcu has been added in 
https://github.com/zephyrproject-rtos/zephyr/issues/65459

MaureenHelm replied:
You don't need a driver; devicetree bindings and corresponding compatible nodes would be sufficient to get the base address of the trgmux from devicetree.

manuargue replied:
then you will need to convert the base address to peripheral hw index, because the HAL is index-based. That can be done with the help of some macros.

sumitbatra-nxp replied:
Done !!

MaureenHelm commented on dts/bindings/sensor/nxp,s32-qdec.yaml:
```diff
@@ -0,0 +1,96 @@
+          <3 TRGMUX_IP_OUTPUT_LCU1_0_INP_I1            TRGMUX_IP_INPUT_SIUL2_IN3>;
+
+  lcu-inst:
+    type: int
```
This should be a phandle

sumitbatra-nxp replied:
Same reply as above

sumitbatra-nxp replied:
done !!

decsny reviewed (dismissed):
You are making a binding and driver called `nxp,s32-qdec` and enabling on the board using the s32k3 chip. But there is no QDEC on the k3 chip. Can you please explain this.

bperseghetti commented:
> You are making a binding and driver called `nxp,s32-qdec` and enabling on the board using the s32k3 chip. But there is no QDEC on the k3 chip. Can you please explain this.

@decsny  I would assume this is not different than naming choices for these, I am not aware of the exact feature portfolios of these but do all "mcux" have qdec or all "stm32"?
- [nxp,mcux-qdec.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/sensor/nxp%2Cmcux-qdec.yaml)

- [atmel,sam-tc-qdec.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/sensor/atmel%2Csam-tc-qdec.yaml)

- [st,stm32-qdec.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/sensor/st%2Cstm32-qdec.yaml)

manuargue commented on west.yml:
```diff
@@ -193,7 +193,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: 1ed023da7f5541b78099b5a030507a2839ce554f
+      revision: pull/297/head
```
the PR in HAL has been merged, please update the revision

sumitbatra-nxp replied:
Now its PR 306

decsny commented:
I misunderstood slightly what is going on here. In this case, the s32 makes sense in the binding name, since it is using a combination of peripherals on the s32. But, it seems like the LCU and TNGMUX be separate nodes in DT?

sumitbatra-nxp commented:
@MaureenHelm .. For - static uint32_t Clockwise_overflow_count, CounterCW_overflow_count;
I have taken a 2D array - emios_overflow_count[emios_inst][emios_hw_ch_cw]
I should be able to operate on the overflow counters per emios inst and channel (programmable from Device Tree) by this now.

I have also made all the variables unique as per device instance

bperseghetti commented on west.yml:
```diff
@@ -193,7 +193,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: 1ed023da7f5541b78099b5a030507a2839ce554f
+      revision: pull/306/head
```
now: 5c3e33b9870b67d245ecb324baa7f7217c4415fe

manuargue commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,429 @@
+												\
+	PINCTRL_DT_INST_DEFINE(n);								\
+	TRGMUX_IP_INIT_CONFIG(n)								\
+	LCU_IP_INIT_CONFIG(n)									\
```
Not a change request on this PR if you plan to address this in a follow up PR, but beware that the instance `n` refers to the QDEC instance (which is a software instance because there is no QDEC hw peripheral), but is also used to configure the hw instance `LCU_NXP_S32_GET_INSTANCE()`. The same applies for TRGMUX. The only way to solve this is to separate TRGMUX/LCU into their own drivers so they can keep their own private data for the respective hw instance as I initially mentioned.

manuargue reviewed (dismissed):
Approving the S32 integration with the TODOs discussed above to be addressed in a follow up pr.

igalloway commented:
Hello @manuargue  @MaureenHelm  - I'm trying to follow this PR as it is blocking some other activity on our side. Are there still unresolved changes that need to be made or has @sumitbatra-nxp  addressed them at this point?

sumitbatra-nxp committed:
drivers: sensor: qdec_s32: Add QDEC support for S32

Add code to configure and program Lcu, Trgmux and Emios_Icu IPs to
get the the rotations by the motor in radians.

Co-authored-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>
Co-authored-by: Peter van der Perk <peter.vanderperk@nxp.com>
Co-authored-by: Mayank Mahajan <mayankmahajan.x@nxp.com>
Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

MaureenHelm commented on drivers/sensor/qdec_nxp_s32/qdec_nxp_s32.c:
```diff
@@ -0,0 +1,444 @@
+}
+
+#define EMIOS_NXP_S32_CALLBACK(n)								\
+	static uint32_t qdec##n##_emios_cw_overflow_count;					\
+	static uint32_t qdec##n##_emios_ccw_overflow_count;					\
+												\
+	static void qdec##n##_emios_overflow_count_cw_callback(void)				\
+	{											\
+		qdec##n##_emios_cw_overflow_count++;						\
+	}											\
+												\
+	static void qdec##n##_emios_overflow_count_ccw_callback(void)				\
+	{											\
+		qdec##n##_emios_ccw_overflow_count++;						\
+	}											\
```
This is I meant in https://github.com/zephyrproject-rtos/zephyr/pull/65035#discussion_r1408479814:
```c
static void qdec_emios_overflow_count_cw_callback(const struct device *dev)
{
	struct qdec_s32_data *data = dev->data;

	data->emios_cw_overflow_count++;
}

#define EMIOS_NXP_S32_CALLBACK(n) \
static void qdec##n##_emios_overflow_count_cw_callback(void)
{ \
	qdec_emios_overflow_count_cw_callback(DEVICE_DT_INST_GET(n)); \
}
```

That way you don't need `SENSOR_DRIVER_API`  or `QDEC_NXP_S32_FETCH` macros, which neither make sense nor even work (you're back to duplicate symbols again). It doesn't make sense to have instance-specific implementations of sensor API functions that take `const struct device *` as an argument.

sumitbatra-nxp replied:
Thanks @MaureenHelm for helping out..
I have made these changes
